src/runtime/virtcontainers/qemu.go

```diff
@@ -1396,31 +1399,68 @@ func (q *qemu) hotplugAddVhostUserBlkDevice(ctx context.Context, vAttr *config.V
 	}()
 
 	driver := "vhost-user-blk-pci"
-	addr, bridge, err := q.arch.addDeviceToBridge(ctx, vAttr.DevID, types.PCI)
-	if err != nil {
-		return err
-	}
 
-	defer func() {
+	machineType := q.HypervisorConfig().HypervisorMachineType
+
+	switch machineType {
+	case QemuVirt:
+		if q.state.PCIeRootPort <= 0 {
+			return fmt.Errorf("Vhost-user-blk device is a PCIe device if machine type is virt. Need to add the PCIe Root Port by setting the pcie_root_port parameter in the configuration for virt")
+		}
+
+		//The addr of a dev is corresponding with device:function for PCIe in qemu which starting from 0
+		//Since the dev is the first and only one on this bus(root port), it should be 0.
+		addr := "00"
+
+		bridgeId := fmt.Sprintf("%s%d", pcieRootPortPrefix, len(drivers.AllPCIeDevs))
+		drivers.AllPCIeDevs[devID] = true
+
+		bridgeQomPath := fmt.Sprintf("%s%s", qomPathPrefix, bridgeId)
+		bridgeSlot, err := q.qomGetSlot(bridgeQomPath)
 		if err != nil {
-			q.arch.removeDeviceFromBridge(vAttr.DevID)
+			return err
 		}
-	}()
 
-	bridgeSlot, err := vcTypes.PciSlotFromInt(bridge.Addr)
-	if err != nil {
-		return err
-	}
-	devSlot, err := vcTypes.PciSlotFromString(addr)
-	if err != nil {
-		return err
-	}
-	vAttr.PCIPath, err = vcTypes.PciPathFromSlots(bridgeSlot, devSlot)
+		devSlot, err := vcTypes.PciSlotFromString(addr)
+		if err != nil {
+			return err
+		}
 
-	if err = q.qmpMonitorCh.qmp.ExecutePCIVhostUserDevAdd(q.qmpMonitorCh.ctx, driver, devID, vAttr.DevID, addr, bridge.ID); err != nil {
-		return err
-	}
+		vAttr.PCIPath, err = vcTypes.PciPathFromSlots(bridgeSlot, devSlot)
+		if err != nil {
+			return err
+		}
+
+		if err = q.qmpMonitorCh.qmp.ExecutePCIVhostUserDevAdd(q.qmpMonitorCh.ctx, driver, devID, vAttr.DevID, addr, bridgeId); err != nil {
+			return err
+		}
+
+	default:
+		addr, bridge, err := q.arch.addDeviceToBridge(ctx, vAttr.DevID, types.PCI)
```

Comment:
Hrm.  So arguably this puts the switch on machine type in the wrong place.  The reason you need this is that despite the name, `addDeviceToBridge` is not a useful cross-machine abstraction, and in fact its interfaces are misdesigned so that it can't really be tweaked into one. A better approach long term would be a proper abstraction for placing PCI devices.  I've looked at doing this, but haven't had the time to attempt it.

I recognize that that's a much bigger refactor than you probably want to take on just to get ARM working.  So, I won't request it be changed just on that basis.  That said, a major reason that the code in this area is such a godawful mess is that people keep adding things without refactoring the base to support it cleanly, so there's that ://

Reply:
Thanks @dgibson -, it's pretty not good to do this here. However, to refactor this is beyond this PR. So, thanks for the "precious" approval!